### PR TITLE
Switch logging from glog to go-kit/log

### PR DESCRIPTION
**Pull request by tdabasinskas**
As described under #30, `glog` package does not support outputting logs as JSON, making it quite difficult to capture logs of JIRAlert in cloud-native deployments.

The PR replaces `glog` with [`go-kit/kit/log`](https://github.com/go-kit/kit/tree/master/log) package, which is used both by [Prometheus](https://github.com/prometheus/prometheus/blob/master/cmd/prometheus/main.go#L36) and [Thanos](https://github.com/improbable-eng/thanos/blob/master/cmd/thanos/main.go#L23). This change introduces additional command line options: `--log.level` and `--log.format` that allow configuring logging verbosity and output format. In addition, all log messages were updated to use structured logging.

### CODE REVIEW

**Comment by tdabasinskas** [comment]
Hi, @free / @bwplotka 
Could you look into this when you have some time?

File: pkg/notify/notify.go [hunk 1]
```diff
@@ -56,45 +57,45 @@ func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
...
 		if resolutionTime.Add(time.Duration(*r.conf.ReopenDuration)).After(time.Now()) {
-			log.Infof("Issue %s for %s was resolved on %s, reopening", issue.Key, issueLabel, resolutionTime.Format(time.RFC3339))
-			return r.reopen(issue.Key)
+			level.Info(logger).Log("msg", "issue was resolved after reopen duration, reopening", "issue", issue.Key, "label", issueLabel, "reopenDuration", *r.conf.ReopenDuration, "resolutionTime", resolutionTime)
+			return r.reopen(issue.Key, logger)
 		}
 	}
 
-	log.Infof("No issue matching %s found, creating new issue", issueLabel)
+	level.Info(logger).Log("msg", "no issue, matching the label, found, opening a new one", "label", issueLabel)
```
Comment (free) [inline pkg/notify/notify.go]:
```suggestion
	level.Info(logger).Log("msg", "No issue matching the label found, creating new issue", "label", issueLabel)
```

File: pkg/notify/notify.go [hunk 2]
```diff
@@ -106,23 +107,23 @@ func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
...
 	}
 
 	if err := r.tmpl.Err(); err != nil {
 		return false, err
 	}
-	retry, err = r.create(issue)
+	retry, err = r.create(issue, logger)
 	if err == nil {
-		log.Infof("Issue created: key=%s ID=%s", issue.Key, issue.ID)
+		level.Info(logger).Log("msg", "issue successfully created", "issue", issue.Key, "issueID", issue.ID)
```
Comment (free) [inline pkg/notify/notify.go]:
```suggestion
		level.Info(logger).Log("msg", "Issue created", "key", issue.Key, "id", issue.ID)
```
or
```suggestion
		level.Info(logger).Log("msg", "Issue created", "issue.Key", issue.Key, "issue.ID", issue.ID)
```

File: pkg/notify/notify.go [hunk 3]
```diff
@@ -172,67 +173,68 @@ func toIssueLabel(groupLabels alertmanager.KV) string {
...
 
-func (r *Receiver) search(project, issueLabel string) (*jira.Issue, bool, error) {
+func (r *Receiver) search(project, issueLabel string, logger log.Logger) (*jira.Issue, bool, error) {
 	query := fmt.Sprintf("project=\"%s\" and labels=%q order by resolutiondate desc", project, issueLabel)
 	options := &jira.SearchOptions{
 		Fields:     []string{"summary", "status", "resolution", "resolutiondate"},
 		MaxResults: 2,
 	}
-	log.V(1).Infof("search: query=%v options=%+v", query, options)
+	level.Debug(logger).Log("msg", "searching for existing issue", "query", query, "options", options)
```
Comment (free) [inline pkg/notify/notify.go]: Please leave these log messages as they were (minus the formatted parameters). They are only intended for debugging, go after one another in the log and are easier to eyeball as they are:
```
msg="Search" query="XXX"
msg="  found" issue="YYY"
```
or
```
msg="Create" issue="XXX"
msg="  done" key="YYY"
```

File: pkg/notify/notify.go [hunk 4]
```diff
@@ -172,67 +173,68 @@ func toIssueLabel(groupLabels alertmanager.KV) string {
...
+func (r *Receiver) reopen(issueKey string, logger log.Logger) (bool, error) {
 	transitions, resp, err := r.client.Issue.GetTransitions(issueKey)
 	if err != nil {
-		return handleJiraError("Issue.GetTransitions", resp, err)
+		return handleJiraError("Issue.GetTransitions", resp, err, logger)
 	}
 	for _, t := range transitions {
 		if t.Name == r.conf.ReopenState {
-			log.V(1).Infof("reopen: issueKey=%v transitionID=%v", issueKey, t.ID)
+			level.Debug(logger).Log("msg", "reopening issue", "issue", issueKey, "transitionID", t.ID)
```
Comment (free) [inline pkg/notify/notify.go]: Same here.

File: pkg/notify/notify.go [hunk 5]
```diff
@@ -172,67 +173,68 @@ func toIssueLabel(groupLabels alertmanager.KV) string {
...
 			return false, nil
 		}
 	}
 	return false, fmt.Errorf("JIRA state %q does not exist or no transition possible for %s", r.conf.ReopenState, issueKey)
 }
 
-func (r *Receiver) create(issue *jira.Issue) (bool, error) {
-	log.V(1).Infof("create: issue=%v", *issue)
+func (r *Receiver) create(issue *jira.Issue, logger log.Logger) (bool, error) {
+	level.Debug(logger).Log("msg", "creating new issue", "issue", *issue)
```
Comment (free) [inline pkg/notify/notify.go]: And here.

File: pkg/notify/notify.go [hunk 6]
```diff
@@ -56,45 +57,45 @@ func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
 		// The set of JIRA status categories is fixed, this is a safe check to make.
 		if issue.Fields.Status.StatusCategory.Key != "done" {
 			// Issue is in a "to do" or "in progress" state, all done here.
-			log.V(1).Infof("Issue %s for %s is unresolved, nothing to do", issue.Key, issueLabel)
+			level.Info(logger).Log("msg", "nothing to do as issue is unresolved", "issue", issue.Key, "label", issueLabel)
```
Comment (free) [inline pkg/notify/notify.go]: And I'd name the key for `issue.Key` either `"key"` or `"issue.Key"`, to separate it from `"issue"`, which is a dump of the  issue object.

Same everywhere else.

File: pkg/template/template.go [hunk 7]
```diff
@@ -48,10 +48,10 @@ func (t *Template) Err() error {
 // Execute parses the provided text (or returns it unchanged if not a Go template), associates it with the templates
 // defined in t.tmpl (so they may be referenced and used) and applies the resulting template to the specified data
 // object, returning the output as a string.
-func (t *Template) Execute(text string, data interface{}) string {
-	log.V(2).Infof("Executing template %q...", text)
+func (t *Template) Execute(text string, data interface{}, logger log.Logger) string {
+	level.Debug(logger).Log("msg", "executing template", "template", text)
 	if !strings.Contains(text, "{{") {
-		log.V(2).Infof("  returning unchanged.")
+		level.Debug(logger).Log("msg", "returning unchanged template")
```
Comment (free) [inline pkg/template/template.go]: Same as above, these log messages are intended for debugging only, they go after one another in the log so they are easier to eyeball if they are simpler and the follow-up messages are indented.

File: cmd/jiralert/main.go [hunk 8]
```diff
@@ -38,64 +43,84 @@ func main() {
...
+		if *logFormat == logFormatJson {
+			logger = log.NewJSONLogger(log.NewSyncWriter(os.Stderr))
+		}
+		logger = level.NewFilter(logger, lvl)
+
+		logger = log.With(logger, "ts", log.DefaultTimestampUTC, "caller", log.DefaultCaller)
+	}
 
-	config, _, err := config.LoadFile(*configFile)
+	level.Info(logger).Log("msg", "starting JIRAlert", "version", Version)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply cmd/jiralert/main.go]: well yes and no (: but the fact it used in *most* of Prometheus related code, makes it the default choice. It is bit tricky, epecially that it:
* fails if have odd number of args
* fails if something is not string (or has no stringer)

But otherwise it's a casual structured logger, better then glog definintely (IMO). The is also rule that those log fields should be lower case and not full sentences (no fullstop). No sure where that rule was written though ;p

File: cmd/jiralert/main.go [hunk 9]
```diff
@@ -112,11 +137,16 @@ func main() {
 		*listenAddress = ":" + os.Getenv("PORT")
 	}
 
-	log.Infof("Listening on %s", *listenAddress)
-	log.Fatal(http.ListenAndServe(*listenAddress, nil))
+	level.Info(logger).Log("msg", "listening", "address", *listenAddress)
+	err = http.ListenAndServe(*listenAddress, nil)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to start HTTP server", "address", *listenAddress)
+		os.Exit(1)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply cmd/jiralert/main.go]: Again, in every other Prometheus project it is `1`. Wonder what's the reason behind 255? Does not exists: https://www.tldp.org/LDP/abs/html/exitcodes.html

File: pkg/notify/notify.go [hunk 10]
```diff
@@ -56,45 +57,45 @@ func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
...
 			return false, nil
 		}
 		if r.conf.WontFixResolution != "" && issue.Fields.Resolution != nil &&
 			issue.Fields.Resolution.Name == r.conf.WontFixResolution {
 			// Issue is resolved as "Won't Fix" or equivalent, log a message just in case.
-			log.Infof("Issue %s for %s is resolved as %q, not reopening", issue.Key, issueLabel, issue.Fields.Resolution.Name)
+			level.Info(logger).Log("msg", "issue is resolved, not reopening", "issue", issue.Key, "label", issueLabel, "resolution", issue.Fields.Resolution.Name)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/notify/notify.go]: Keeping `"Issue was` -> `"issue was` might be the standard here.

File: cmd/jiralert/main.go [hunk 11]
```diff
@@ -13,20 +13,25 @@ import (
...
 	unknownReceiver = "<unknown>"
+	logFormatLogfmt = "logfmt"
+	logFormatJson = "json"
 )
 
 var (
 	listenAddress = flag.String("listen-address", ":9097", "The address to listen on for HTTP requests.")
 	configFile    = flag.String("config", "config/jiralert.yml", "The JIRAlert configuration file")
+	logLevel      = flag.String("log.level", "info", "Log filtering level")
+	logFormat     = flag.String("log.format", logFormatLogfmt, "Log format to use")
```
[the remark this answers is not in the document]
Comment (tdabasinskas) [reply cmd/jiralert/main.go]: Makes sense. Added.

File: cmd/jiralert/main.go [hunk 12]
```diff
@@ -38,64 +43,84 @@ func main() {
...
+			lvl = level.AllowWarn()
+		case "debug":
+			lvl = level.AllowDebug()
+		case "info":
+		default:
+			lvl = level.AllowInfo()
+		}
+		logger = log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+		if *logFormat == logFormatJson {
+			logger = log.NewJSONLogger(log.NewSyncWriter(os.Stderr))
```
[the remark this answers is not in the document]
Comment (tdabasinskas) [reply cmd/jiralert/main.go]: Good catch!

File: cmd/jiralert/main.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tdabasinskas) [reply cmd/jiralert/main.go]: `255` looks a bit strange to me, as I don't recall seeing it in any other GoLang projects. But that's up to you, guys.

File: cmd/jiralert/main.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tdabasinskas) [reply cmd/jiralert/main.go]: I was considering using capitalized string within `msg`, but looking into [the Prometheus code](https://github.com/prometheus/prometheus/blob/master/storage/remote/wal_watcher.go#L142), it seems to be lowercase. Since most other parameters are going to be lowercase, I think it make sense to keep `msg` lowercase as well.

File: cmd/jiralert/main.go [hunk 13]
```diff
@@ -112,11 +137,16 @@ func main() {
...
 
-	log.Infof("Listening on %s", *listenAddress)
-	log.Fatal(http.ListenAndServe(*listenAddress, nil))
+	level.Info(logger).Log("msg", "listening", "address", *listenAddress)
+	err = http.ListenAndServe(*listenAddress, nil)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to start HTTP server", "address", *listenAddress)
+		os.Exit(1)
+	}
+
```
[the remark this answers is not in the document]
Comment (tdabasinskas) [reply cmd/jiralert/main.go]: Removed

File: pkg/notify/notify.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tdabasinskas) [reply pkg/notify/notify.go]: I've changed it to `issue is resolved, not reopening`

File: pkg/notify/notify.go [hunk 14]
```diff
@@ -56,45 +57,45 @@ func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
...
 			return false, nil
 		}
 
 		resolutionTime := time.Time(issue.Fields.Resolutiondate)
 		if resolutionTime.Add(time.Duration(*r.conf.ReopenDuration)).After(time.Now()) {
-			log.Infof("Issue %s for %s was resolved on %s, reopening", issue.Key, issueLabel, resolutionTime.Format(time.RFC3339))
-			return r.reopen(issue.Key)
+			level.Info(logger).Log("msg", "issue was resolved after reopen duration, reopening", "issue", issue.Key, "label", issueLabel, "reopenDuration", *r.conf.ReopenDuration, "resolutionTime", resolutionTime)
```
[the remark this answers is not in the document]
Comment (tdabasinskas) [reply pkg/notify/notify.go]: Agree and adding RFC-3339 format back.

File: pkg/notify/notify.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tdabasinskas) [reply pkg/notify/notify.go]: I've updated all messages there to `Doing YYY, as  XXX`. Hope it OK

File: pkg/notify/notify.go (hunk 3 above)
Comment (tdabasinskas) [reply pkg/notify/notify.go]: Personally, I'm not sure I like this extra formatting when talking about logs. Nevertheless, I've change all occurrences back as requested.

File: cmd/jiralert/main.go [hunk 15]
```diff
@@ -38,64 +43,84 @@ func main() {
...
-	// Override -alsologtostderr default value.
-	if alsoLogToStderr := flag.Lookup("alsologtostderr"); alsoLogToStderr != nil {
-		alsoLogToStderr.DefValue = "true"
-		alsoLogToStderr.Value.Set("true")
-	}
 	flag.Parse()
 
-	log.Infof("Starting JIRAlert version %s", Version)
+	var logger log.Logger
+	{
```
[the remark this answers is not in the document]
Comment (tdabasinskas) [reply cmd/jiralert/main.go]: Moved to `setupLogger()`

**Comment by tdabasinskas** [comment]
Thanks for the comments. I've tried addressing all of them, but let me know if I've missed something or if further changes are required.

File: pkg/notify/notify.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (free) [reply pkg/notify/notify.go]: With apologies for walking back my request, I think stating the cause before the outcome makes more sense. I also find the new log messages somewhat verbose. I'm normally quite careful with grammar, but log files should be free of fluff.

File: pkg/notify/notify.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (free) [reply pkg/notify/notify.go]: Ping.

File: pkg/notify/notify.go [hunk 16]
```diff
@@ -56,45 +57,45 @@ func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
...
 		if resolutionTime.Add(time.Duration(*r.conf.ReopenDuration)).After(time.Now()) {
-			log.Infof("Issue %s for %s was resolved on %s, reopening", issue.Key, issueLabel, resolutionTime.Format(time.RFC3339))
-			return r.reopen(issue.Key)
+			level.Info(logger).Log("msg", "reopening issue, as it was resolved after reopen duration, ", "key", issue.Key, "label", issueLabel, "reopenDuration", *r.conf.ReopenDuration, "resolutionTime", resolutionTime.Format(time.RFC3339))
+			return r.reopen(issue.Key, logger)
 		}
 	}
 
-	log.Infof("No issue matching %s found, creating new issue", issueLabel)
+	level.Info(logger).Log("msg", "opening a new issue, as no issue matching the label was found, ", "label", issueLabel)
```
Comment (free) [inline pkg/notify/notify.go]: Not strictly related to your change, but since we now have a `reopen_duration` this log message should say `"no recent matching issue found, creating new issue".
```suggestion
	level.Info(logger).Log("msg", "no recent matching issue found, creating new issue", "label", issueLabel)
```

File: pkg/notify/notify.go [hunk 17]
```diff
@@ -172,67 +173,68 @@ func toIssueLabel(groupLabels alertmanager.KV) string {
...
 	if err != nil {
-		retry, err := handleJiraError("Issue.Search", resp, err)
+		retry, err := handleJiraError("Issue.Search", resp, err, logger)
 		return nil, retry, err
 	}
 	if len(issues) > 0 {
 		if len(issues) > 1 {
 			// Swallow it, but log a message.
-			log.Infof("More than one issue matched %s, will only update last issue: %+v", query, issues)
+			level.Warn(logger).Log("msg", "more than one issue matched, updating only the last one", "query", query, "issues", issues)
```
Comment (free) [inline pkg/notify/notify.go]: I was going to suggest rewording this, for the same reason as above: we now have a `reopen_duration`, which may mean we don't actually update the most recent issue.

But the `reopen_duration` is all but guaranteed to result in multiple issues with the same label being created over time, so I'd say this warning should be dropped entirely. Or on second (now third) thought, demoted to debug. And slightly reworded.
```suggestion
			level.Warn(logger).Log("msg", "more than one issue matched, picking most recently resolved", "query", query, "issues", issues)
```

File: pkg/notify/notify.go (hunk 3 above)
Comment (free) [reply pkg/notify/notify.go]: You seem to have left out the indentation, though. That's the thing that allows you to be brief in all but the first message.

Comment (tdabasinskas) [reply pkg/notify/notify.go]: Could you point to the exact lines where indentation is still missing? I thought I've added it (two extra spaces) to the places I've removed it initially. It's possible I've missed some places though.

**Comment by tdabasinskas** [comment]
Hello, @free ,

Sorry for the delay. Could you please let me know if any further changes are still required? I've committed your suggestions and the only two remaining comments, according to GitHub, are on outdated files.

File: pkg/notify/notify.go (hunk 3 above)
Comment (free) [reply pkg/notify/notify.go]: Seems like indentation is all OK now.

File: pkg/template/template.go [hunk 18]
```diff
@@ -65,12 +65,12 @@ func (t *Template) Execute(text string, data interface{}) string {
 	}
 	tmpl, t.err = tmpl.New("").Parse(text)
 	if t.err != nil {
-		log.V(2).Infof("  parse failed.")
+		level.Warn(logger).Log("msg", "failed to parse template")
```
Comment (free) [inline pkg/template/template.go]:
```suggestion
		level.Warn(logger).Log("msg", "failed to parse template", "template", text)
```

File: pkg/template/template.go [hunk 19]
```diff
@@ -65,12 +65,12 @@ func (t *Template) Execute(text string, data interface{}) string {
...
 		return ""
 	}
 	var buf bytes.Buffer
 	t.err = tmpl.Execute(&buf, data)
 	ret := buf.String()
-	log.V(2).Infof("  returning %q.", ret)
+	level.Debug(logger).Log("msg", "returning parsed template", "template", ret)
```
Comment (free) [inline pkg/template/template.go]:
```suggestion
	level.Debug(logger).Log("msg", "  template output", "output", ret)
```

File: pkg/notify/notify.go [hunk 20]
```diff
@@ -172,67 +173,68 @@ func toIssueLabel(groupLabels alertmanager.KV) string {
...
 	if err != nil {
-		retry, err := handleJiraError("Issue.Search", resp, err)
+		retry, err := handleJiraError("Issue.Search", resp, err, logger)
 		return nil, retry, err
 	}
 	if len(issues) > 0 {
 		if len(issues) > 1 {
 			// Swallow it, but log a message.
-			log.Infof("More than one issue matched %s, will only update last issue: %+v", query, issues)
+			level.Debug(logger).Log("msg", "more than one issue matched, picking most recently resolved", "query", query, "issues", issues)
```
Comment (free) [inline pkg/notify/notify.go]:
```suggestion
			level.Debug(logger).Log("msg", "  more than one issue matched, picking most recently resolved", "query", query, "issues", issues)
```

File: cmd/jiralert/main.go [hunk 21]
```diff
@@ -38,64 +43,62 @@ func main() {
...
 		}
 
 		conf := config.ReceiverByName(data.Receiver)
 		if conf == nil {
-			errorHandler(w, http.StatusNotFound, fmt.Errorf("receiver missing: %s", data.Receiver), unknownReceiver, &data)
+			errorHandler(w, http.StatusNotFound, fmt.Errorf("receiver missing: %s", data.Receiver), unknownReceiver, &data, logger)
 			return
 		}
-		log.V(1).Infof("Matched receiver: %q", conf.Name)
+		level.Debug(logger).Log("msg", "matched receiver", "receiver", conf.Name)
```
Comment (free) [inline cmd/jiralert/main.go]:
```suggestion
		level.Debug(logger).Log("msg", "  matched receiver", "receiver", conf.Name)
```

File: pkg/config/config.go [hunk 22]
```diff
@@ -55,19 +55,19 @@ func LoadFile(filename string) (*Config, []byte, error) {
...
 // with a given base directory.
-func resolveFilepaths(baseDir string, cfg *Config) {
+func resolveFilepaths(baseDir string, cfg *Config, logger log.Logger) {
 	join := func(fp string) string {
 		if len(fp) == 0 || filepath.IsAbs(fp) {
 			return fp
 		}
 		absFp := filepath.Join(baseDir, fp)
-		log.V(2).Infof("Relative path %q resolved to %q", fp, absFp)
+		level.Debug(logger).Log("msg", "relative configuration path resolved", "relativePath", fp, "absolutePath", absFp)
```
Comment (free) [inline pkg/config/config.go]:
```suggestion
		level.Debug(logger).Log("msg", "resolved relative configuration path", "relativePath", fp, "absolutePath", absFp)
```

File: pkg/notify/notify.go [hunk 23]
```diff
@@ -56,45 +57,45 @@ func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
 		// The set of JIRA status categories is fixed, this is a safe check to make.
 		if issue.Fields.Status.StatusCategory.Key != "done" {
 			// Issue is in a "to do" or "in progress" state, all done here.
-			log.V(1).Infof("Issue %s for %s is unresolved, nothing to do", issue.Key, issueLabel)
+			level.Debug(logger).Log("msg", "doing nothing, as issue is unresolved", "key", issue.Key, "label", issueLabel)
```
Comment (free) [inline pkg/notify/notify.go]:
```suggestion
			level.Debug(logger).Log("msg", "issue is unresolved, nothing to do", "key", issue.Key, "label", issueLabel)
```

File: pkg/notify/notify.go [hunk 24]
```diff
@@ -56,45 +57,45 @@ func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
...
 			return false, nil
 		}
 		if r.conf.WontFixResolution != "" && issue.Fields.Resolution != nil &&
 			issue.Fields.Resolution.Name == r.conf.WontFixResolution {
 			// Issue is resolved as "Won't Fix" or equivalent, log a message just in case.
-			log.Infof("Issue %s for %s is resolved as %q, not reopening", issue.Key, issueLabel, issue.Fields.Resolution.Name)
+			level.Info(logger).Log("msg", "not reopening issue, as it was resolved", "key", issue.Key, "label", issueLabel, "resolution", issue.Fields.Resolution.Name)
```
Comment (free) [inline pkg/notify/notify.go]:
```suggestion
			level.Info(logger).Log("msg", "issue was resolved as won't fix, not reopening", "key", issue.Key, "label", issueLabel, "resolution", issue.Fields.Resolution.Name)
```

File: pkg/notify/notify.go [hunk 25]
```diff
@@ -56,45 +57,45 @@ func (r *Receiver) Notify(data *alertmanager.Data) (bool, error) {
...
 			return false, nil
 		}
 
 		resolutionTime := time.Time(issue.Fields.Resolutiondate)
 		if resolutionTime.Add(time.Duration(*r.conf.ReopenDuration)).After(time.Now()) {
-			log.Infof("Issue %s for %s was resolved on %s, reopening", issue.Key, issueLabel, resolutionTime.Format(time.RFC3339))
-			return r.reopen(issue.Key)
+			level.Info(logger).Log("msg", "reopening issue, as it was resolved after reopen duration, ", "key", issue.Key, "label", issueLabel, "reopenDuration", *r.conf.ReopenDuration, "resolutionTime", resolutionTime.Format(time.RFC3339))
```
Comment (free) [inline pkg/notify/notify.go]:
```suggestion
			level.Info(logger).Log("msg", "issue was recently resolved, reopening", "key", issue.Key, "label", issueLabel, "resolution_time", resolutionTime.Format(time.RFC3339), "reopen_duration", *r.conf.ReopenDuration)
```
